test/FsAutoComplete.Tests.Lsp/Program.fs

```diff
@@ -76,14 +77,63 @@ let tests =
 [<EntryPoint>]
 let main args =
   let outputTemplate = "[{Timestamp:HH:mm:ss} {Level:u3}] [{SourceContext}] {Message:lj}{NewLine}{Exception}"
-  let verbose = args |> Seq.contains "--debug"
-  let switch = LoggingLevelSwitch()
 
-  if verbose
-  then
-    switch.MinimumLevel <- LogEventLevel.Verbose
-  else
-    switch.MinimumLevel <- LogEventLevel.Information
+  let parseLogLevel args =
+    let debugMarker = "--debug"
+    let logMarker = "--log="
+    let logLevel =
+      if args |> Array.contains "--debug" then
+        Logging.LogLevel.Verbose
+      else
+        match args |> Array.tryFind (fun arg -> arg.StartsWith logMarker) |> Option.map (fun log -> log.Substring(logMarker.Length)) with
+        | Some ("warn" | "warning") ->
+          Logging.LogLevel.Warn
+        | Some "error" ->
+          Logging.LogLevel.Error
+        | Some "fatal" ->
+          Logging.LogLevel.Fatal
+        | Some "info" ->
+          Logging.LogLevel.Info
+        | Some "verbose" ->
+          Logging.LogLevel.Verbose
+        | Some "debug" ->
+          Logging.LogLevel.Debug
+        | _ -> Logging.LogLevel.Info
+    let args =
+      args 
+      |> Array.except [debugMarker]
+      |> Array.filter (fun arg -> not <| arg.StartsWith logMarker)
+    logLevel, args
+  let expectoToSerilogLevel =
+    function
+    | Logging.LogLevel.Debug -> LogEventLevel.Debug
+    | Logging.LogLevel.Verbose -> LogEventLevel.Verbose
+    | Logging.LogLevel.Info -> LogEventLevel.Information
+    | Logging.LogLevel.Warn -> LogEventLevel.Warning
+    | Logging.LogLevel.Error -> LogEventLevel.Error
+    | Logging.LogLevel.Fatal -> LogEventLevel.Fatal
+  let parseLogExcludes (args: string[]) =
+    let excludeMarker = "--exclude-from-log="
+    let toExclude =
+      args
+      |> Array.filter (fun arg -> arg.StartsWith excludeMarker)
+      |> Array.collect (fun arg -> arg.Substring(excludeMarker.Length).Split(','))
+    let args =
+      args
+      |> Array.filter (fun arg -> not <| arg.StartsWith excludeMarker)
+    toExclude, args
+
+  let logLevel, args = parseLogLevel args
+  let switch = LoggingLevelSwitch(expectoToSerilogLevel logLevel)
```

Comment:
by the way, if you have any ideas around more granular runtime control of logging levels for log categories, please let me know! it sucks only being able to change this at startup.

Reply:
2 notes:

* great analysis overall, I agree that we could use a more...holistic approach to error logging. at the same time, we need to not log seemingly normal situations (like when the editor calls for decorations but we have no typecheck results available for whatever reason) at a verbosity where normal users can see them. we see a lot of logs where people misunderstand that specific error, for example.
* I think this is the first time I've seen a footnote in github comments, it's quite nice :D